File: algorithms/Java/stacks/Redundant_Parenthesis.java
import java.util.Stack;

public class Redundant_Parenthesis {
    // Function to check redundant brackets in a balanced expression
    public static boolean checkRedundancy(String s) {
		// create a stack of characters
		Stack<Character> st = new Stack<>();
		char[] str = s.toCharArray();
		// Iterate through the given expression
		for (char ch : str) {
			// if current character is close parenthesis ')'
			if (ch == ')') {
				char top = st.peek();
				st.pop();
				// If immediate pop have open parenthesis '('
				// duplicate brackets found
				boolean flag = true;
				while (top != '(') {
					// Check for operators in expression
					if (top == '+' || top == '-'
							|| top == '*' || top == '/') {
						flag = false;
					}
					// Fetch top element of stack
					top = st.peek();
					st.pop();
				}

				// If operators not found
				if (flag == true) {
					return true;
				}
			} else {
				st.push(ch); // push open parenthesis '(',
			}			 // operators and operands to stack
		}
		return false;
	}
    public static void main(String[] args) {
        String str = "((a+b)*(b+c))";
        if(checkRedundancy(str)){
            System.out.println("The given expression contains redundant parenthesis.");
        }else{
            System.out.println("The given expression does not contains redundant parenthesis.");
        }
    }
}


// Input :
// For example:
// 1. ((a+b)) 
// 2. (a+b*(c-d)) 

// Output:
// 1. True, ((a+b)) can reduced to (a+b), this is Redundant
// 2. False, (a+b*(c-d)) doesn't have any redundant or multiple
// brackets

File: algorithms/CPlusPlus/Stacks/redundant-parenthesis.cpp
//  RedundantParenthesis
// Given a string of balanced expression, find if it contains a redundant parenthesis or not. A set of parenthesis are redundant if the same sub-expression is surrounded by unnecessary or multiple brackets. Print True if redundant, else False.

// Algorithm
// 1. We iterate through the given expression and for each character in the expression, if the character is an open parenthesis ‘(‘ or any operators, we push it to the stack.
// 2. If the character is close parenthesis ‘)’, then pop characters from the stack till matching open parenthesis ‘(‘ is found. 
// For any sub-expression of expression, if we are able to pick any sub-expression of expression surrounded by (), then we again left with () as part of string, we have redundant braces. 
// We iterate through the given expression and for each character in the expression, if the character is an open parenthesis ‘(‘ or any of the operators or operands, we push it to the stack. If the character is close parenthesis ‘)’, then pop characters from the stack till matching open parenthesis ‘(‘ is found. 
// Now for redundancy two condition will arise while popping- 
// 1. If immediate pop hits an open parenthesis ‘(‘, then we have found a duplicate parenthesis. For example, (((a+b))+c) has duplicate brackets around a+b. When we reach the second “)” after a+b, we have “((” in the stack. Since the top of stack is an opening bracket, we conclude that there are duplicate brackets. 
// 2. If immediate pop doesn’t hit any operand(‘*’, ‘+’, ‘/’, ‘-‘) then it indicates the presence of unwanted brackets surrounded by expression. For instance, (a)+b contain unwanted () around a thus it is redundant. 



// solution
#include <bits/stdc++.h>
using namespace std;

int main()
{
    cout << "Enter the string:" << endl;
    string s;
    cin >> s;
    // create a stack of characters
    stack<char> st;
    bool ans = false;

    // Iterate through the given expression
    for (int i = 0; i < s.size(); i++)
    {
        if (s[i] == '+' or s[i] == '-' or s[i] == '*' or s[i] == '/')
        {
            st.push(s[i]);
        }
        else if (s[i] == '(')
        {
            // if current character is close parenthesis '('
            st.push(s[i]);
        }
        else if (s[i] == ')')
        {
            // if current character is close parenthesis ')'
            if (st.top() == '(')
            {
                ans = true;
            }
            while (st.top() == '+' or st.top() == '-' or st.top() == '*' or st.top() == '/')
            {
                st.pop();
            }
            st.pop();
        }
    }

    if (ans)
    {
        cout << "True";
    }
    else
    {
        cout << "False";
    }
}


// Input :
// For example:
// 1. ((a+b))
// 2. (a+b*(c-d))

// Output:
// 1. True, ((a+b)) can reduced to (a+b), this is Redundant
// 2. False, (a+b*(c-d)) doesn't have any redundant or multiple
// brackets
